internal/tracefs/kprobe.go

```diff
@@ -0,0 +1,381 @@
+package tracefs
+
+import (
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/sys"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+var (
+	ErrInvalidInput = errors.New("invalid input")
+
+	ErrInvalidMaxActive = errors.New("can only set maxactive on kretprobes")
+)
+
+func KprobeCheckLite(symbol string, pid int) error {
+	fd, err := tryAttach(symbol, pid)
+	if err != nil {
+		return err
+	}
+	_ = fd.Close()
+	return nil
+}
+
+func tryAttach(symbol string, pid int) (*sys.FD, error) {
+	args := ProbeArgs{
+		Symbol: symbol,
+		Pid:    pid,
+	}
+
+	// Use tracefs if kprobe PMU is missing.
+	fd, _, _, err := TracefsProbe(KprobeType, args)
+	return fd, err
+}
+
+// TracefsProbe creates a trace event by writing an entry to <tracefs>/[k,u]probe_events.
+// A new trace event group name is generated on every call to support creating
+// multiple trace events for the same kernel or userspace symbol.
+// Path and offset are only set in the case of uprobe(s) and are used to set
+// the executable/library path on the filesystem and the offset where the probe is inserted.
+// A perf event is then opened on the newly-created trace event and returned to the caller.
+func TracefsProbe(typ ProbeType, args ProbeArgs) (*sys.FD, uint64, string, error) {
```

Comment:
This is the only real change from this PR, all the rest is mostly moving things around, but this basically extract the fd creation to `internal/tracefs` and keeps the `perfEvent struct` creation in `link`

Reply:
I think this is nice, and I might even go a step further: remove `OpenTracepointPerfEvent` from this function and instead of `sys.FD` return a new `type Probe struct{...}` which has a `Close()` method that calls `removeTraceFSProbeEvent`. Maybe even a finalizer, but that is TBD. I think that would make the code easier to follow and gets rid of the `pe.tracefsID != 0` checks in `perfEvent.Close`.

Nit: how about calling this `tracefs.NewProbe` instead?